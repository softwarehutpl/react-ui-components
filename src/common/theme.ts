const theme: { [key: string]: { [key: string]: { [key: string]: string } } } = {
  colors: {
    primary: {
      base: '#000000',
      light: '#ffffff',
      dark: '#a8b3bd',
    },
    secondary: {
      base: '#3690e3',
      light: '#75b0e6',
      dark: '#034d91',
    },
    error: {
      base: '#f44336',
      light: '#f56c62',
      dark: '#a1140a',
    },
    warning: {
      base: '#faa72d',
      light: '#ffc36b',
      dark: '#995b00',
    },
    info: {
      base: '#7979d4',
      light: '#9797d1',
      dark: '#4c4cc7',
    },
    success: {
      base: '#4caf50',
      light: '#88c28a',
      dark: '#124f14',
    },
    disabled: {
      base: '#d9d9d9',
      light: '#f0eded',
      dark: '#8c8989',
    },
  },
  // TODO: change it in dedicated task for that (change ease-out for example and add more options)
  transitions: {
    all: {
      fast: 'all 0.5s ease-out',
      mid: 'all 2s ease-out',
      slow: 'all 4s ease-out',
    },
    background: {
      fast: 'background-color 0.5s ease-out',
      mid: 'background-color 2s ease-out',
      slow: 'background-color 4s ease-out',
    },
    fontColor: {
      fast: 'color 0.5s ease-out',
      mid: 'color 2s ease-out',
      slow: 'color 4s ease-out',
    }
  }
};

export default theme;
