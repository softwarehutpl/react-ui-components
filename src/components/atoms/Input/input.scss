@import 'src/common/styles/variables.scss';

.test-class {
  color: $color-black;
  background-color: $color-white;
  border-color: $color-cerise;
}