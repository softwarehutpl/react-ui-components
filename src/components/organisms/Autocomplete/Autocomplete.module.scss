@import 'src/common/styles/variables.scss';

.InputContainer {
  width: 100%;
}

.Input {
  width: 100%;
}