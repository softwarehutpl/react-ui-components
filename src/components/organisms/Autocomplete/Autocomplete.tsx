import React from 'react';

export const Autocomplete: React.StatelessComponent<{}> = () => (
  <div className="">Autocomplete
  </div>
);
