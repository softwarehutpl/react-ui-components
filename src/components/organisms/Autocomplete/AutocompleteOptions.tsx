import React from 'react';
import styled, { css } from 'styled-components';

export interface OptionItemProps {
  isActive?: boolean;
  isHover?: boolean;
  color?: string;
  hoverOption?: number;
};

export interface OptionWrapperProps {
  width?: number;
  height?: number;
  padding?: number;
  margin?: number;
}

export interface AutoCompleteOptionsProps extends OptionWrapperProps, OptionItemProps {
  options: string[];
  onClick?: (e: React.MouseEvent<HTMLLIElement>) => void;
  onMouseEnter?: any;
  onMouseLeave?:(e: React.MouseEvent<HTMLLIElement>) => void;
  activeOption?: number;
  hoverOption?: number;
};

export const OptionListWrapper = styled.ul<OptionWrapperProps>`
  width: ${({ width }) => (width ? `${width}px` : '100%' )};
  height: ${({ height }) => (height ? `${height}px` : 'auto' )};
  overflow: auto;
  box-shadow: 0px 0px 10px rgba(0,0,0,0.12);
  border-radius: 15px;
  padding: ${({ padding }) => `${padding}px` };
  margin:${({ margin }) => `${margin}px` };
`;

export const OptionItem = styled.li<OptionItemProps>`
  list-style-type: none;
  padding: 15px;
  background-color: ${({ color, theme, isActive }) => ((isActive && color) && theme.colors[color].dark)};
  color: ${({ theme, color }) => (color && theme.colors[color].base)};
  cursor: pointer;

  ${({ isHover, color }) => css`
    background-color: ${({ theme }) => ((color && isHover ) && theme.colors[color].dark)};
  `}
`;

const AutocompleteOptions: React.FunctionComponent<AutoCompleteOptionsProps> = ({
  options,
  onClick,
  onMouseEnter,
  onMouseLeave,
  activeOption,
  color,
  width,
  height,
  padding,
  margin,
  hoverOption,
}) => (
  <OptionListWrapper
    height={height}
    width={width}
    padding={padding}
    margin={margin}
  >
    {options.map((option: string, index: number) => (
      <OptionItem
        key={option}
        onClick={onClick}
        onMouseEnter={(e) => onMouseEnter(e, index)}
        onMouseLeave={onMouseLeave}
        isActive={activeOption === index}
        color={color}
        isHover={hoverOption === index}
      >
        {option}
      </OptionItem>
    ))}
  </OptionListWrapper>
);

AutocompleteOptions.defaultProps = {
  padding: 0,
  margin: 0,
};

export default AutocompleteOptions;
