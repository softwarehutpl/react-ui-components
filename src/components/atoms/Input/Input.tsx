import React from 'react';
import styled from 'styled-components';
import './input.scss';

type Props = {
  buttonTitle?: string;
  onChange: () => void;
  color?: string;
  className?: string;
  placeholder?: string;
  disabled?: boolean;
  fontColor?: string;
  noBorder?: boolean;
  borderColor?: string;
  hoverBorderColor?: string;
  focusBorderColor?: string;
  borderWidth?: number;
  borderRadius?: number;
  backgroundColor?: string;
  hoverBackgroundColor?: string;
  focusBackgroundColor?: string;
  transitionEffect?: string;
  showPlaceholderOnFocus?: boolean;
  margin?: number;
  padding?: number;
}

const Input = (
  {
    onChange,
    className,
    placeholder,
    disabled,
  }: Props
) => (
  <input
    className={className}
    onChange={onChange}
    disabled={disabled}
    placeholder={placeholder}
  />
);

const defaultProps = {
  color: 'primary',
  fontColor: '#000',
  noBorder: false,
  disabled: false,
  className: '',
  borderWidth: 2,
  margin: 0,
  padding: 15,
  borderRadius: 0,
  borderColor: '#000',
  showPlaceholderOnFocus: false,
};

Input.defaultProps = defaultProps;

export default styled(Input)`${({ 
  theme,
  color = defaultProps.color,
  backgroundColor,
  fontColor = defaultProps.fontColor,
  noBorder = defaultProps.noBorder,
  borderColor = defaultProps.borderColor,
  borderWidth = defaultProps.borderWidth,
  borderRadius = defaultProps.borderRadius,
  margin = defaultProps.margin,
  padding = defaultProps.padding,
  hoverBackgroundColor,
  hoverBorderColor,
  focusBackgroundColor,
  focusBorderColor,
  transitionEffect,
  showPlaceholderOnFocus,
}) => ({
  'background-color': backgroundColor || theme.colors[color].light,
  'color': fontColor || theme.colors[color].dark,
  'border': noBorder ? 'none' : 'solid',
  'border-color': borderColor || theme.colors[color].dark,
  'border-width': `${borderWidth}px`,
  'border-radius': `${borderRadius}px`,
  'margin': `${margin}px`,
  'padding': `${padding}px`,
  '&:hover': {
    'cursor': 'pointer',
    'background-color': hoverBackgroundColor || focusBackgroundColor || theme.colors[color].light,
    'border-color': hoverBorderColor || focusBorderColor || theme.colors[color].dark,
  },
  '&:focus': {
    'cursor': 'pointer',
    'background-color': focusBackgroundColor || theme.colors[color].light,
    'border-color': focusBorderColor || theme.colors[color].dark,
  },
  '&:focus::-webkit-input-placeholder': showPlaceholderOnFocus ? {
    'font-size': '.75em',
    'position': 'relative',
    'top': '-15px',
    'transition': transitionEffect ? theme.transitions.all[transitionEffect] : '0.2s ease-out',
  } : '',
  '&::-webkit-input-placeholder': showPlaceholderOnFocus ? {
    'transition': transitionEffect ? theme.transitions.all[transitionEffect] : '0.2s ease-out',
  } : '',
  'outline': 'none',
  'transition': transitionEffect ? theme.transitions.all[transitionEffect] : 'none',
  '&[disabled]': {
    'background-color': theme.colors.disabled.base,
    'border-color': theme.colors.disabled.dark,
    'color': '#a6a6a6',
    '&:hover': {
      'cursor': 'auto',
    },
  },
})}`
