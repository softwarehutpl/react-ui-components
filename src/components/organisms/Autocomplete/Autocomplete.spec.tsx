import React from 'react';
import { create, act } from 'react-test-renderer';
import { ThemeProvider } from 'styled-components';
import 'jest-styled-components';
import Autocomplete from './Autocomplete';
import Input from '../../atoms/Input/Input';
import AutocompleteOptions from './AutocompleteOptions';
import theme from '../../../common/theme';

const options = [
  'Papaya',
  'Persimmon',
  'Paw Paw',
  'Prickly Pear',
  'Peach',
  'Pomegranate',
  'Pineapple'
];

describe('<Autocomplete />', () => {
  it('should render corectly', () => {
    const component = create(
      <ThemeProvider theme={theme}>
        <Autocomplete options={options} />
      </ThemeProvider>).toJSON();
    expect(component).toMatchSnapshot();
  });

  it('should set value on change event', () => {
    const event = { currentTarget: {
      value: 'test1'
    }} as any;

    const component = create(
      <ThemeProvider theme={theme}>
        <Autocomplete options={options} />
      </ThemeProvider>
    );
    const instance = component.root;
    const input = instance.findByType('input');
    expect(input.props.value).toEqual('')

    act(() => {
      input.props.onChange(event);
    });

    expect(input.props.value).toEqual('test1');
  });

  it('should call onSelect prop', () => {
    const mockOnSelect= jest.fn();
    const component = create(
      <ThemeProvider theme={theme}>
        <Autocomplete options={options} onSelect={mockOnSelect} />
      </ThemeProvider>);

    const instance = component.root;
    const autocomplete = instance.findByType(Autocomplete);

    act(() => {
      autocomplete.props.onSelect();
    });
    expect(mockOnSelect.mock.calls.length).toEqual(1);
  });

  it('should call list item onClick method', () => {
    const event = {
      currentTarget: {
        innerText: 'test1'
      }
    };
    const mockOnClick = jest.fn();
    const component = create( <AutocompleteOptions onClick={mockOnClick} options={options} />);
    const instance = component.root;
    const list = instance.findByType(AutocompleteOptions);

    act(() => {
      list.props.onClick(event);
    });

    expect(mockOnClick.mock.calls.length).toEqual(1);
  });

  it('should call list item onMouseEnter method', () => {
    const mockOnEnter = jest.fn();
    const component = create( <AutocompleteOptions onMouseEnter={mockOnEnter} options={options} />);
    const instance = component.root;
    const list = instance.findByType(AutocompleteOptions);

    act(() => {
      list.props.onMouseEnter();
    });
    
    expect(mockOnEnter.mock.calls.length).toEqual(1);
  });

  it('should call list item onKeyDown method', () => {
    const mockOnKeyDown = jest.fn();
    const component = create( <Input onKeyDown={mockOnKeyDown} color="primary" theme={theme} />);
    const instance = component.root;
    const input = instance.findByType(Input);

    act(() => {
      input.props.onKeyDown();
    });
    
    expect(mockOnKeyDown.mock.calls.length).toEqual(1);
  });

  it('should render with propery hover option' , () => {
    const component = create( 
      <AutocompleteOptions 
        options={options} 
        hoverOption={1}
      />
    );

    const instance = component.root;
    const list = instance.findByType(AutocompleteOptions);
    expect(list.props.hoverOption).toEqual(1);
  });

  it('should render with proper activeOption on onKeyDown event', () => {
    const rootComponent = create(
      <ThemeProvider theme={theme}>
        <Autocomplete options={options} />
      </ThemeProvider>
    );
    const optionsComponent = create( 
      <AutocompleteOptions 
        options={options} 
        activeOption={0}
      />
    );

    const input = rootComponent.root.findByType('input');;
    const list = optionsComponent.root.findByType(AutocompleteOptions);

    act(() => {
      input.props.onKeyDown({keyCode: 13});
    });
    expect(list.props.activeOption).toEqual(0);
  });
});
