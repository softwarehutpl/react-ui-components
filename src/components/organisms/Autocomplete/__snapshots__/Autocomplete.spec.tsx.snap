// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Autocomplete /> should render corectly 1`] = `
.c2 {
  background-color: #FFF;
  color: #000;
  border: solid;
  border-color: #f6f6f6;
  border-width: 1px;
  border-radius: 5px;
  margin: 0px;
  padding: 15px;
  width: 200px;
  outline: none;
  -webkit-transition: none;
  transition: none;
}

.c2:hover {
  border-color: #f6f6f6;
  background-color: #ffffff;
  border-width: 2px;
}

.c2:focus {
  border-width: 2px;
  border-color: #f6f6f6;
  background-color: #ffffff;
}

.c2:focus::-webkit-input-placeholder {
  font-size: .75em;
  display: none;
  position: relative;
  top: -15px;
  -webkit-transition: 0.2s ease-out;
  transition: 0.2s ease-out;
}

.c2::-webkit-input-placeholder {
  -webkit-transition: 0.2s ease-out;
  transition: 0.2s ease-out;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
  width: 200px;
}

.c0 {
  width: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

<div
  className="c0"
>
  <div
    className="c1 InputContainer"
    direction="column"
    width={200}
  >
    <input
      className="c2 Input"
      color="primary"
      direction="column"
      disabled={false}
      onChange={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      placeholder="Search"
      value=""
      width={200}
    />
  </div>
</div>
`;
